src/languages/r.js

```diff
@@ -18,13 +18,20 @@ export default function(hljs) {
   // handled in a separate mode. See `test/markup/r/names.txt` for examples.
   // FIXME: Support Unicode identifiers.
   const IDENT_RE = /(?:(?:[a-zA-Z]|\.[._a-zA-Z])[._a-zA-Z0-9]*)|\.(?!\d)/;
-  const SIMPLE_IDENT = /[a-zA-Z][a-zA-Z_0-9]*/;
+  const NUMBER_TYPES = regex.either(
+    // Special case: only hexadecimal binary powers can contain fractions
+    /0[xX][0-9a-fA-F]+\.[0-9a-fA-F]*[pP][+-]?\d+i?/,
+    // Hexadecimal numbers without fraction and optional binary power
+    /0[xX][0-9a-fA-F]+(?:[pP][+-]?\d+)?[Li]?/,
+    // Decimal numbers
+    /(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?[Li]?/
+  );
+  const OPERATORS = /[=!<>:]=|\|\||&&|:::?|<-|<<-|->>|->|\|>|[-+*\/?!$&|:<=>@^~]|\*\*/;
+  const PUNCTUATION = /\[\[|[(){}[\]\\,]/;
```

Comment:
This just makes my head hurt, can we use regex either and have an array of punctuation?  For some reason the operators is a bit easier to read.

Reply:
Like this? 252483ca4a0e1cd407b5787d9deb470d1ce5dece